### DISCUSSION TITLE
How can I enforce https?

### DISCUSSION THREAD

**Discussion by maxschmeling** [vercel/next.js#16856] [Help] [answered]
I'm hosting on Heroku and want to always serve over HTTPS. The only place I see this mentioned in docs / GitHub / discussions / web search is #355 from 2016 which says you have to use a custom server which the docs talk about as something to be avoided.

It feels like I'm missing something because I would imagine this is an *extremely* common need. I'm guessing the Vercel hosting platform handles this automatically so the framework doesn't have to, so is the only option for non Vercel hosting to create a custom server and handle the redirect there?

**Comment by johansundwall** [comment, marked as answer]
Heroku doesn't offer a built in way to do this.
https://help.heroku.com/J2R1S4T8/can-heroku-force-an-application-to-use-ssl-tls

> Redirects need to be performed at the application level as the Heroku router does not provide this functionality. You should code the redirect logic into your application.

So you will need to do it on application level. The only way to do that is with a custom server.
If you know that you will continue to host your application on Heroku or similar there is nothing wrong with using a custom server.

**Reply by karianpour** [reply]
I looks like that there is something wrong with using customer server as it says : 
"A custom server will remove important performance optimizations, like serverless functions and Automatic Static Optimization."

Do you thinks that it is ok to use custom server?
